src/signedTransaction.ts

```diff
@@ -0,0 +1,171 @@
+import {
+  Encodable,
+  encodeMsgpack,
+  decodeMsgpack,
+} from './encoding/encoding.js';
+import { Address } from './encoding/address.js';
+import { Transaction } from './transaction.js';
+import { LogicSig } from './logicsig.js';
+import {
+  EncodedMultisig,
+  encodedMultiSigToEncodingData,
+  encodedMultiSigFromEncodingData,
+  ENCODED_MULTISIG_SCHEMA,
+} from './types/transactions/index.js';
+import {
+  AddressSchema,
+  FixedLengthByteArraySchema,
+  NamedMapSchema,
+} from './encoding/schema/index.js';
+
+export class SignedTransaction implements Encodable {
+  static encodingSchema = new NamedMapSchema([
+    {
+      key: 'txn',
+      valueSchema: Transaction.encodingSchema,
+      required: true,
+      omitEmpty: true,
+    },
+    {
+      key: 'sig',
+      valueSchema: new FixedLengthByteArraySchema(64),
+      required: false,
+      omitEmpty: true,
+    },
+    {
+      key: 'msig',
+      valueSchema: ENCODED_MULTISIG_SCHEMA,
+      required: false,
+      omitEmpty: true,
+    },
+    {
+      key: 'lsig',
+      valueSchema: LogicSig.encodingSchema,
+      required: false,
+      omitEmpty: true,
+    },
+    {
+      key: 'sgnr',
+      valueSchema: new AddressSchema(),
+      required: false,
+      omitEmpty: true,
+    },
+  ]);
+
+  /**
+   * The transaction that was signed
+   */
+  public readonly txn: Transaction;
+
+  /**
+   * Transaction signature
+   */
+  public readonly sig?: Uint8Array;
+
+  /**
+   * Multisig structure
+   */
+  public readonly msig?: EncodedMultisig;
+
+  /**
+   * Logic signature
+   */
+  public readonly lsig?: LogicSig;
+
+  /**
+   * The signer, if signing with a different key than the Transaction type `sender` property indicates
+   */
+  public readonly sgnr?: Address;
+
+  constructor({
+    txn,
+    sig,
+    msig,
+    lsig,
+    sgnr,
+  }: {
+    txn: Transaction;
+    sig?: Uint8Array;
+    msig?: EncodedMultisig;
+    lsig?: LogicSig;
+    sgnr?: Address;
+  }) {
+    this.txn = txn;
+    this.sig = sig;
+    this.msig = msig;
+    this.lsig = lsig;
+    this.sgnr = sgnr;
+
+    let numberOfSigs = 0;
+    if (sig) numberOfSigs += 1;
+    if (msig) numberOfSigs += 1;
+    if (lsig) numberOfSigs += 1;
+    if (numberOfSigs > 1) {
+      throw new Error(
+        `SignedTransaction must not have more than 1 signature. Got ${numberOfSigs}`
+      );
+    }
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  public getEncodingSchema() {
+    return SignedTransaction.encodingSchema;
+  }
+
+  public toEncodingData(): Map<string, unknown> {
+    const data = new Map<string, unknown>([
+      ['txn', this.txn.toEncodingData()],
+      ['sig', this.sig],
+      ['sgnr', this.sgnr],
+    ]);
+    if (this.msig) {
+      data.set('msig', encodedMultiSigToEncodingData(this.msig));
```

Comment:
wouldn't it retain the empty `sig` property if msig provided?

Reply:
If `msig` is defined then `sig` must be `undefined`. Undefined values are [dropped during msgpack](https://github.com/algorand/js-algorand-sdk/blob/b864c24ac3007a2bba08f0bad1c98d1617891ebe/src/encoding/schema/map.ts#L62-L64) and JSON preparation, so it makes no difference.

A previous version of the code looked something like:
```javascript
if (this.sig) {
    data.set('sig', this.sig)
}
if (this.msig) {
    data.set('msig', encodedMultiSigToEncodingData(this.msig))
}
```
Due to dropping undefined values this is equivalent to the current code.